### ISSUE TITLE
Integration test / sample project

### ISSUE THREAD

**Issue by Jarod42** [premake/premake-core#1502]
**What's your question?**

I began to create integration test (see https://github.com/premake/premake-core/pull/1500) (and spot bugs in codelite generator)
but PR was not the right place for the needed discussion.

Do you think it is a good idea?
if yes, how to organize them (which directories, ci-scripts, ...)
For sample projects, I would say the "minimum" number needed to test the features.

How to handle "known bugs", not (yet) supported feature (in generator).
People updating sample tests don't necessary know how to fix each generator.
Probably true also for people working on core

How to handle target action limitation (some IDE might not support some features (so generators cannot improved/fixed))
I think premake/generator should warn about unsupported feature.
filter by feature might help to write "portable" testing projects.

**Comment by starkos** [comment]
> Generators have little logic to test (there are mainly a dump of premake configuration in correct format).

Oh how very wrong you are :)

I get the idea, you want something along the lines of [snapshot testing](https://jestjs.io/docs/en/snapshot-testing). But there are so many possible combinations of settings and switches across all the tools, to test effectively you are going to end up with a _massive_ set of tests which will take quite a long time to run. Someone will have to curate, manage, and maintain them—if that's even possible at that scale. And each time something changes, someone will have to manually review the (potentially large number of) failures to identify the culprit(s). I don't see how we're gaining enough value to justify to cost.

>>Is there something you're trying to test that isn't possible with the current test framework?
> In which order generate links and linkoptions (see #1492):

I don't see why this can't be unit tested? Yes, you have to write a separate test for each generator, but…eh? If there is something that prevents unit testing, that should be fixed.

I appreciate you taking the time and effort to think it through and write it up, but at this time I don't think this is a feasible approach for this project. But if you think you're on to something, feel free to prove me wrong! I'm going to mark this issue "closed as answered" for now, but I'll continue to get notified if you add more comments.

**Comment by Jarod42** [comment]
> > Generators have little logic to test (there are mainly a dump of premake configuration in correct format).

> Oh how very wrong you are :)

I meant premake allows to factorize common logic in core.

When I wrote custom generator (to generate compile_command.json for clang tooling)
My errors were to miss to handle some features (force_include, sysincludedirs) rather than the "logic" (output "cc" for c files and "cxx" for c++ files).

When I quickly look at codelite generator, I don't really see logic neither.

With sample projects, I would have my tests for free.

> But there are so many possible combinations of settings and switches across all the tools, to test effectively you are going to end up with a massive set of tests which will take quite a long time to run.

I don't ask for any combinations, but I think with some sample projects, we might cover the big part.

I hesitate on modularity of projects, but for example, something like:

- project to test compile options (includes (force, sys, pch, regular), defines, buildoptions).
- project to test links options (kind, links, linksoptions)
- project to test per file configuration
- project to test toolset choice
- dotnet project

I already know that codelite fails for the first two (sysincludedir, and kind=sharedLib) (PR coming soon :) )

I also know there are some features that are hard and not really possible to test with integration testing (as optimization settings) or might go too further (Do objects have symbols?)

I see premake.lua as portable (I avoid specific rule depending of generator except target directories to avoid conflict) way to generate project via the generators.
and there is nothing to test that.

I think there is mostly 2 kind of new feature in premake:
- core ones with new api (named flag as "Optimize", per file configuration, ...)
- addition of generator (codeblock, cmake, compile_command.json, ...)

people doing the former doesn't necessary know each generator.
people working on generator which might have limited knowledge of core

The current way to test an API is to test it for each generator (which is not completely under your control/repository BTW with the plugin system).

Tests of new generators (such as code::block) are not trivial to test as we have (in addition to possible logical test relative to generator) to list premake api for what I would call the dump/conversion of premake config into the generated "solution". 

If we go that way, we should know status (truly supported, workaround (treat sysincludedirs as regular includedirs) or partially implemented, not yet implemented, unsupported)  of features by generator.

That would be useful for tests, but also as documentation

> I don't see why this can't be unit tested?

Maybe partially.
It reminds me a project which mocks (path) system calls in UTs, but most bugs happens as mock doesn't behave as system (max path limitation, type of error/exception returned...). The current UTs expect an output which is never really used.

Additionally, in UTs I see, they rely on implementation details with no ways to know if it is important or not:
- `sysincludedirs` before or after `includedirs` or no matter? (it does actually in UTs, but would not for toolsets I know)
- `linksoptions` before or after `links` or no matter (or broken either ways)? Links order is important (at least for gcc) (So I think it is currently broken either way :/ ). from Uts point of view, it is similar to above.

**Comment by starkos** [comment]
> With sample projects, I would have my tests for free.

Free for you, not free for the folks who have to maintain them long term. 

If you genuinely want this in core, you'll have to put together a more compelling example than what you've shown so far. Everything you've asked to test is already being testing on other generators. If it is difficult to do for CodeLite, then the code should be reworked to follow the example set by the Visual Studio generator.

**Comment by Jarod42** [comment]
I began to create some (in separate repo this time): https://github.com/Jarod42/premake-sample-projects

(I dislike the way I test them though as I placed testing code directly in github action, but it is a detail from now
and I fallback to premake4 for apt-get install).

Whereas in sample project, we know the feature is (at least partially) working. :)

Is it true that then, we have to dissociated premake's bug from bugs of the tools
(Just discovered that premake4+codelite had issue with semi colon in MACRO definition, not yet investigate if it is a escaping issue from premake4, or issue from codelite)

I recently made a fix (fix quote escaping in Codelite) with UT, the question was "is the UT correct?".
That question won't happen with projects ;-)

> If it is difficult to do for CodeLite

I don't think it is related to Codelite.

Other generator should answer question like:
How should we handle `defines('MACRO="with space quote \" and semi ;"')`

and UTs for that is just our expectation. It has to be run in real by the tools to check our assumption.

I will have a look at VS.

> If it is difficult to do for CodeLite

As I said, when a new feature is added in core, ideally all generators should be updated.
I suspect that Codelite is less used/tested than the other more "popular".
I even think sad that simple projects were broken.

Hard for contributors to handle each generator, but I think create sample projects would be possible and allow people knowing the generator to fix the issues.
